File: admin/src/main/java/com/beat/admin/promotion/application/dto/response/CarouselHandleAllResponse.java
package com.beat.admin.promotion.application.dto.response;

import java.util.List;
import java.util.stream.Collectors;

import com.beat.admin.promotion.application.dto.result.AdminPromotionResults;
import com.beat.admin.promotion.application.dto.result.AdminPromotionResults.AdminPromotionResult;
import com.fasterxml.jackson.annotation.JsonProperty;

public record CarouselHandleAllResponse(
	@JsonProperty("modifiedPromotions")
	List<PromotionResponse> modifiedPromotionResponses
) {

	private static CarouselHandleAllResponse fromResponses(List<PromotionResponse> modifiedPromotionResponses) {
		return new CarouselHandleAllResponse(modifiedPromotionResponses);
	}

	public static CarouselHandleAllResponse from(AdminPromotionResults promotionResults) {
		List<PromotionResponse> modifiedPromotionResponses = promotionResults.promotionResults().stream()
			.map(PromotionResponse::from)
			.collect(Collectors.toList());
		return CarouselHandleAllResponse.fromResponses(modifiedPromotionResponses);
	}

	public record PromotionResponse(
		Long promotionId,
		String newImageUrl,
		boolean isExternal,
		String redirectUrl,
		String carouselNumber
	) {
		private static PromotionResponse of(Long promotionId, String newImageUrl, boolean isExternal,
			String redirectUrl, String carouselNumber) {
			return new PromotionResponse(promotionId, newImageUrl, isExternal, redirectUrl, carouselNumber);
		}

		public static PromotionResponse from(AdminPromotionResult promotionResult) {
			return PromotionResponse.of(
				promotionResult.promotionId(),
				promotionResult.newImageUrl(),
				promotionResult.isExternal(),
				promotionResult.redirectUrl(),
				promotionResult.carouselNumber()
			);
		}
	}
}

File: apps/admin/src/main/kotlin/com/beat/apps/admin/promotion/api/response/CarouselHandleAllResponse.kt
package com.beat.apps.admin.promotion.api.response

import com.beat.application.admin.promotion.AdminPromotionResults
import com.beat.application.admin.promotion.AdminPromotionResults.AdminPromotionResult
import com.fasterxml.jackson.annotation.JsonProperty

data class CarouselHandleAllResponse(
    @get:JsonProperty("modifiedPromotions") val modifiedPromotionResponses: List<PromotionResponse>
) {
    constructor(
        promotionResults: AdminPromotionResults
    ) : this(promotionResults.promotionResults.map { PromotionResponse(it) })

    data class PromotionResponse(
        val promotionId: Long?,
        val newImageUrl: String,
        val isExternal: Boolean,
        val redirectUrl: String,
        val carouselNumber: String,
    ) {
        constructor(
            promotionResult: AdminPromotionResult
        ) : this(
            promotionId = promotionResult.promotionId,
            newImageUrl = promotionResult.newImageUrl,
            isExternal = promotionResult.isExternal,
            redirectUrl = promotionResult.redirectUrl,
            carouselNumber = promotionResult.carouselNumber,
        )
    }
}
